src/pudl/transform/eia861.py

```diff
@@ -175,14 +434,176 @@
     ("WA", "Wahkiakurn", "Wahkiakum"),
 ], columns=["state", "eia_county", "fips_county"])
 
+BA_NAME_FIXES = pd.DataFrame([
+    ("Omaha Public Power District", 14127, "OPPD"),
+    ("Kansas City Power & Light Co", 10000, "KCPL"),
+    ("Toledo Edison Co", 18997, pd.NA),
+    ("Ohio Edison Co", 13998, pd.NA),
+    ("Cleveland Electric Illum Co", 3755, pd.NA),
+], columns=["balancing_authority_name_eia",
+            "balancing_authority_id_eia",
+            "balancing_authority_code_eia",
+            ]
+)
+
+CUSTOMER_CLASSES = [
+    "commercial",
+    "industrial",
+    "other",
+    "residential",
+    "total",
+    "transportation"
+]
+
+###############################################################################
+# EIA Form 861 Transform Helper functions
+###############################################################################
+
+
+def _filter_customer_cols(df, customer_classes):
+    regex = f"^({'_|'.join(customer_classes)}).*$"
+    return df.filter(regex=regex)
+
+
+def _filter_non_customer_cols(df, customer_classes):
+    regex = f"^(?!({'_|'.join(customer_classes)})).*$"
+    return df.filter(regex=regex)
+
+
+def _ba_code_backfill(df):
+    """
+    Backfill Balancing Authority Codes based on codes in later years.
+
+    Note:
+        The BA Code to ID mapping can change from year to year. If a Balancing Authority
+        is bought by another entity, the code may change, but the old EIA BA ID will be
+        retained.
+
+    Args:
+        ba_eia861 (pandas.DataFrame): The transformed EIA 861 Balancing
+            Authority dataframe (balancing_authority_eia861).
+
+    Returns:
+        pandas.DataFrame: The balancing_authority_eia861 dataframe, but with
+        many fewer NA values in the balancing_authority_code_eia column.
+
+    """
+    start_len = len(df)
+    start_nas = len(df.loc[df.balancing_authority_code_eia.isnull()])
+    logger.info(
+        f"Started with {start_nas} missing BA Codes out of {start_len} "
+        f"records ({start_nas/start_len:.2%})")
+    ba_ids = (
+        df[["balancing_authority_id_eia", "balancing_authority_code_eia", "report_date"]]
+        .drop_duplicates()
+        .sort_values(["balancing_authority_id_eia", "report_date"])
+    )
+    ba_ids["ba_code_filled"] = (
+        ba_ids.groupby("balancing_authority_id_eia")[
+            "balancing_authority_code_eia"]
+        .apply(lambda x: x.bfill())
+    )
+    ba_eia861_filled = df.merge(ba_ids, how="left")
+    ba_eia861_filled = (
+        ba_eia861_filled.assign(
+            balancing_authority_code_eia=lambda x: x.ba_code_filled)
+        .drop("ba_code_filled", axis="columns")
+    )
+    end_len = len(ba_eia861_filled)
+    if start_len != end_len:
+        raise AssertionError(
+            f"Number of rows in the dataframe changed {start_len}!={end_len}!"
+        )
```

Comment:
i wish there was just some built-in way to do this all of the time.

Reply:
Some people do this kind of checking in their data processing pipelines with decorators (which wrap a function and allow you look at what's going in as well as what's coming out).